### ISSUE TITLE
[5.0.0] new version giving preflight errors

### ISSUE THREAD

**Issue by DRoet** [fastify/fastify-cors#102]
## 🐛 Bug Report

I upgraded from `4.1.0` to `5.0.0` and now I'm getting errors in development (localhost). I can't really figure out what changed based on the release notes

## To Reproduce

```js
fastify.register(require('fastify-cors'), {
   origin: new RegExp('http://localhost'),
   credentials: true,
   optionsSuccessStatus: 200
})
```

Chrome throws the following error and returns a 401 status:

```
Access to XMLHttpRequest at 'http://localhost:3000/api/login' from origin 'http://localhost:8080' has been blocked by CORS policy:
Response to preflight request doesn't pass access control check: It does not have HTTP ok status.
```

I tried turning off preflight/strictPreFlight but no luck there.

## Expected behavior

Local development still works just like in `4.1.0`

## Your Environment

- *node version*: 14
- *fastify version*: >= 3.8.0
- *os*: Docker (linux)
- *any other relevant information*

**Comment by zekth** [comment]
your regex is not good, use:
```js
new RegExp(/http:\/\/localhost/)
```

**Comment by DRoet** [comment]
RegExp should handle the escaping correctly, no?

```js
console.log(new RegExp('http://localhost'))
// prints:  /http:\/\/localhost/
```

Which should turn it into: `origin: /http:\/\/localhost/`

**Comment by zekth** [comment]
you wrote:
```js
   origin: new RegExp(/http://localhost/),
// not 
   origin: new RegExp('http://localhost'),
```

**Comment by DRoet** [comment]
oh my bad, that must have been an copy/paste error for the reproduction on my end. I actually use an ENV variable to fill the RegExp. I edited the example above

**Comment by mcollina** [comment]
@codyzu ptal

**Comment by codyzu** [comment]
I can't reproduce this on my machine with fastify-cors `5.0.0`.

Can you copy the request and response headers for the `OPTIONS` request in the chrome dev tools network tab (right-click on the `OPTIONS` request and copy the request/response headers)?

fastify-cors does not return `401`, so I would be suspicious of the authentication plugin. What are you using for auth?

**Comment by DRoet** [comment]
Hmm you are right, I found the issue in my own `preHandler` hook that does the authentication. I was using `req.context.config` to figure out which route is being called. 
This now differs in version `5.0.0` (maybe intentionally? should I be using a different Fastify method to get this?)

4.1.0:
```js
fastify.addHook('preHandler', (req, reply, done) => {
   console.log(req.context.config)
   // { url: '/api/login', method: 'POST' }
})
```

5.0.0:
```js
fastify.addHook('preHandler', (req, reply, done) => {
   console.log(req.context.config)
   // { url: '*', method: 'OPTIONS' }
})
```

**Comment by codyzu** [comment]
`req.routerPath` and `req.routerMethod` would be the [official](https://www.fastify.io/docs/latest/Request) way to check which router is handling the route (in place of `req.context`).

The problem you are most likely seeing is that in fastify-cors 5.0.0 processing of a preflight request is done in a route: `OPTIONS *` and you should not enforce authentication on that route. Try adding this to your `preHandler` hook:
```js
fastify.addHook('preHandler', (req, reply, done) => {
   // Don't authenticate preflight requests
   if(req.routerPath === '*' && req.routerMethod === 'OPTIONS') {
      return done();
   }

   // Your authentication code ...
})
```

It could also be resolved by not having a global `preHandler` hook for authentication and instead authenticating only the routes or plugins that require authentication.

@mcollina I believe this is a side effect of moving the preflight handling into a route (previously the response would have been handled in fastify-cors `onRequest` hook). Consumers need to be careful not to enforce authentication on the preflight route... Do you see this as a problem? I can add an example in the docs :thinking:

**Comment by mcollina** [comment]
In theory, a cors OPTIONS request should include all headers needed for auth/what would be included in the original request. It should be ok to perform auth for those requests.. could you do a quick check with a browser for both jwt and cookie based solutions?

**Comment by DRoet** [comment]
thanks for the pointer to `req.routerPath`, I've added the preflight skip to the code on my end for now

**Comment by codyzu** [comment]
For simple CORS requests (requests that don't require pre-flighting), the headers are included.

For preflight requests (the CORS OPTIONS request), the browser removes the original request headers. Instead, it includes the name of the headers in `Access-Control-Request-Headers` header, but not the values. See [here](https://www.w3.org/TR/2020/SPSD-cors-20200602/#preflight-request).

I tested this with JWT, but not yet cookies. In any case, it seems like it will continue to create problems for anyone who authenticates all requests.

**Comment by codyzu** [comment]
@DRoet this should be resolved with the 5.1.0 release of fastify-cors, returning the pre 5.0.0 behavior and removing the need to not authenticate OPTIONS requests. With the default options, fastify-cors should now automatically reply to preflight requests before the authentication plugin.

AFAIK this can be closed.

**Comment by DRoet** [comment]
thanks!
